tortoise/verifying_tortoise.go

```diff
@@ -0,0 +1,574 @@
+	t.logger.With().Debug("Initializing genesis layer for verifying tortoise.", genesisLayer.Index(), genesisLayer.Hash().Field("hash"))
+	t.BlocksToBlocks[genesisLayer.Index()] = make(map[types.BlockID]Opinion)
+	for _, blk := range genesisLayer.Blocks() {
+		id := blk.ID()
+		t.BlocksToBlocks[genesisLayer.Index()][blk.ID()] = Opinion{
+			BlocksOpinion: make(map[types.BlockID]vec),
+		}
+		t.GoodBlocksIndex[id] = struct{}{}
+	}
+	t.Last = genesisLayer.Index()
+	t.Evict = genesisLayer.Index()
+	t.Verified = genesisLayer.Index()
+}
+
+// evict makes sure we only keep a window of the last hdist layers.
+func (t *turtle) evict() {
+	// Don't evict before we've Verified more than hdist
+	// TODO: fix potential leak when we can't verify but keep receiving layers
+
+	if t.Verified <= types.GetEffectiveGenesis()+t.Hdist {
+		return
+	}
+	// The window is the last [Verified - hdist] layers.
+	window := t.Verified - t.Hdist
+	t.logger.Info("Window starts %v", window)
+	// evict from last evicted to the beginning of our window.
+	for lyr := t.Evict; lyr < window; lyr++ {
+		t.logger.Info("removing lyr %v", lyr)
+		for blk := range t.BlocksToBlocks[lyr] {
+			delete(t.GoodBlocksIndex, blk)
+		}
+		delete(t.BlocksToBlocks, lyr)
+		t.logger.Debug("evict block %v from maps ", lyr)
+
+	}
+	t.Evict = window
+}
+
+func removeOpinion(o []Opinion, i int) []Opinion {
+	return append(o[:i], o[i+1:]...)
+}
+
+// TODO: cache but somehow check for changes (hare that didn't finish finishes..) maybe check hash?
+func (t *turtle) getSingleInputVectorFromDB(lyrid types.LayerID, blockid types.BlockID) (vec, error) {
+	if lyrid <= types.GetEffectiveGenesis() {
+		return support, nil
+	}
+
+	input, err := t.bdp.GetLayerInputVector(lyrid)
+	if err != nil {
+		return abstain, err
+	}
+
+	str := "["
+	for i, b := range input {
+		str += b.String()
+		if i != len(input)-1 {
+			str += ","
+		}
+	}
+	str += "]"
+	t.logger.With().Debug("got input vector from db", lyrid, log.FieldNamed("query_block", blockid), log.String("input", str))
+
+	for _, bl := range input {
+		if bl == blockid {
+			return support, nil
+		}
+	}
+
+	return against, nil
+}
+
+func (t *turtle) inputVectorForLayer(lyrBlocks []types.BlockID, input []types.BlockID) map[types.BlockID]vec {
+	lyrResult := make(map[types.BlockID]vec, len(lyrBlocks))
+	//XXX : input vector must be pointer so we can differentiate
+	// no support votes to no results at all.
+	if input == nil {
+		// hare didn't finish hence we don't have Opinion
+		// TODO: get hare Opinion when hare finishes
+		for _, b := range lyrBlocks {
+			lyrResult[b] = abstain
+		}
+		return lyrResult
+	}
+
+	for _, b := range input {
+		lyrResult[b] = support
+	}
+
+	for _, b := range lyrBlocks {
+		if _, ok := lyrResult[b]; !ok {
+			lyrResult[b] = against
+		}
+	}
+	return lyrResult
+}
+
+func (t *turtle) BaseBlock() (types.BlockID, [][]types.BlockID, error) {
+	// Try to find a block counting only good blocks
+	for i := t.Last; i > t.Last-t.Hdist; i-- {
```

Comment:
last cant be 0?

Reply:
no really, last is set to `EffectiveGenesis` at the beginning.